Sources/StreamChat/StateLayer/ChannelListState.swift

```diff
@@ -0,0 +1,45 @@
+//
+// Copyright © 2024 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+/// Represents a list of channels matching to the specified query.
+@available(iOS 13.0, *)
+public final class ChannelListState: ObservableObject {
+    private let observer: Observer
+    private let clientConfig: ChatClientConfig
+    let query: ChannelListQuery
+    
+    init(channels: [ChatChannel], query: ChannelListQuery, clientConfig: ChatClientConfig, database: DatabaseContainer) {
+        self.channels = channels
+        self.clientConfig = clientConfig
+        hasLoadedAllChannels = channels.count < query.pagination.pageSize
+        observer = Observer(query: query, chatClientConfig: clientConfig, database: database)
+        self.query = query
+        
+        observer.start(
+            with: .init(channelsDidChange: { [weak self] channels in await self?.setValue(channels, for: \.channels) })
+        )
+    }
+    
+    /// An array of channels for the specified ``ChannelListQuery``.
+    @Published public private(set) var channels = [ChatChannel]()
+    
+    /// True, if all the channels were loaded with paginated loading.
+    @Published public private(set) var hasLoadedAllChannels = false
```

Comment:
@nuno-vieira was asking about jumping to mid page in [here](https://github.com/GetStream/stream-chat-swift/pull/3057/files#r1514736176). 
Since pagination for channels only supports offset/cursor based pagination then I think it is OK if we go for just one boolean state (although I could misuse the API when I start using the cursor based pagination).

For messages pagination we also provided an is loading property, not certain if we should do the same here. `ChatChannelListController` also did not have is loading property.

Reply:
We did discuss something similar with @martinmitrevski when adding watchers property and the decision was to skip such property for now. Makes sense if I remove it from here as well and let the UI side to handle that state instead.